src/App.tsx

```
@@ -1,26 +1,184 @@
+
+      setTodos(prevTodos =>
+        prevTodos.map(todo =>
+          todo.id === updatedTodo.id ? updatedTodo : todo,
+        ),
+      );
+    } catch {
+      setError('Unable to update a todo');
+      setTimeout(() => setError(''), 3000);
+    }
+  };
+
+  const handleMarkAllAsCompleted = (): void => {
+    setTodos(prevTodos =>
+      prevTodos.map(todo => ({ ...todo, completed: true })),
+    );
+  };
+
+  const clearCompleted = async (): Promise<void> => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    try {
+      await Promise.all(
+        completedTodos.map(todo =>
+          deleteTodo(todo.id).catch(() =>
+            setError('Unable to delete some todos'),
+          ),
+        ),
+      );
+      setTodos(prevTodos => prevTodos.filter(todo => !todo.completed));
+    } catch {
+      setError('Error clearing completed todos');
+      setTimeout(() => setError(''), 3000);
+    }
+  };
+
+  const filteredTodos = todos.filter(todo => {
+    switch (filter) {
+      case Filter.Active:
+        return !todo.completed;
+      case Filter.Completed:
+        return todo.completed;
+      default:
+        return true;
+    }
+  });
+
+  const areAllCompleted =
+    todos.length > 0 && todos.every(todo => todo.completed);
+
+  if (!userId) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      {isLoading && (
+        <div className="todo-loader" data-cy="TodoLoader">
+          <span>Loading...</span>
+        </div>
+      )}
+
+      {error && (
```

This component should not be rendered conditionally. Instead use hidden class when you want to hide error message

---

Also move Error layout to separate component